Simplify momentum initialization in example input files

NeilZaim opened the pull request:
Very often, we use in the example input file `particles.momentum_distribution_type= "Gaussian"` when the particles are initialized at rest or with constant momentum. Most of the time, this is done using the (implicit) fact that the mean and thermal velocities are all 0 by default. While, this is not a serious issue, it would be more convenient to see immediately that the particles are initialized with constant momentum or at rest, without having to look if `particles.ux_m` or `particles.ux_th` are defined someplace else in the input file.

So in this PR, I tried to make the initialization of the momentum in our example input files more explicit. More specifically, I have made the following changes:
- I've added a new `momentum_distribution_type`, called `at_rest` (let me know if you'd prefer another name), which just initializes all the momentum components to 0, and I've used it whenever appropriate in the input file.
- Whenever a constant but nonzero initial momentum is used, I've replaced `gaussian` with `constant` in the input file.
- In the documentation, it was written that the parameters determining the initial momentum distributions (`ux_m`, `ux_th`, etc.) were required, which is not the case. I've changed that and added the default values.
- Somewhat unrelated but I've added some `_rt` in the `PlasmaInjector` class.

NeilZaim commented:
It looks like the test `BTD_ReducedSliceDiag` has crashed, I'm not sure why (it didn't crash before merging with development). I will merge again and see if it crashes again.

EZoni commented on Source/Initialization/PlasmaInjector.cpp:
```diff
@@ -477,32 +478,40 @@ void PlasmaInjector::parseDensity (ParmParse& pp)
-        Real uy = 0.;
-        Real uz = 0.;
+    if (mom_dist_s == "at_rest") {
+        constexpr Real ux = 0._rt;
+        constexpr Real uy = 0._rt;
+        constexpr Real uz = 0._rt;
```
```suggestion
        constexpr amrex::Real ux = 0._rt;
        constexpr amrex::Real uy = 0._rt;
        constexpr amrex::Real uz = 0._rt;
```

EZoni commented on Source/Initialization/PlasmaInjector.cpp:
```diff
@@ -477,32 +478,40 @@ void PlasmaInjector::parseDensity (ParmParse& pp)
+        // Construct InjectorMomentum with InjectorMomentumConstant.
+        h_inj_mom.reset(new InjectorMomentum((InjectorMomentumConstant*)nullptr, ux, uy, uz));
+    } else if (mom_dist_s == "constant") {
+        Real ux = 0._rt;
+        Real uy = 0._rt;
+        Real uz = 0._rt;
```
```suggestion
        amrex::Real ux = 0._rt;
        amrex::Real uy = 0._rt;
        amrex::Real uz = 0._rt;
```

EZoni commented on Source/Initialization/PlasmaInjector.cpp:
```diff
@@ -477,32 +478,40 @@ void PlasmaInjector::parseDensity (ParmParse& pp)
-        Real ux_th = 0.;
-        Real uy_th = 0.;
-        Real uz_th = 0.;
+        Real ux_m = 0._rt;
+        Real uy_m = 0._rt;
+        Real uz_m = 0._rt;
+        Real ux_th = 0._rt;
+        Real uy_th = 0._rt;
+        Real uz_th = 0._rt;
```
```suggestion
        amrex::Real ux_m = 0._rt;
        amrex::Real uy_m = 0._rt;
        amrex::Real uz_m = 0._rt;
        amrex::Real ux_th = 0._rt;
        amrex::Real uy_th = 0._rt;
        amrex::Real uz_th = 0._rt;
```

EZoni commented on Source/Initialization/PlasmaInjector.cpp:
```diff
@@ -515,12 +524,12 @@ void PlasmaInjector::parseMomentum (ParmParse& pp)
-        Real ux_th = 0.;
-        Real uy_th = 0.;
-        Real uz_th = 0.;
+        Real ux_m = 0._rt;
+        Real uy_m = 0._rt;
+        Real uz_m = 0._rt;
+        Real ux_th = 0._rt;
+        Real uy_th = 0._rt;
+        Real uz_th = 0._rt;
```
```suggestion
        amrex::Real ux_m = 0._rt;
        amrex::Real uy_m = 0._rt;
        amrex::Real uz_m = 0._rt;
        amrex::Real ux_th = 0._rt;
        amrex::Real uy_th = 0._rt;
        amrex::Real uz_th = 0._rt;
```

EZoni commented on Source/Initialization/PlasmaInjector.cpp:
```diff
@@ -532,8 +541,8 @@ void PlasmaInjector::parseMomentum (ParmParse& pp)
     } else if (mom_dist_s == "maxwell_boltzmann"){
-        Real beta = 0.;
-        Real theta = 10.;
+        Real beta = 0._rt;
+        Real theta = 10._rt;
```
```suggestion
        amrex::Real beta = 0._rt;
        amrex::Real theta = 10._rt;
```

EZoni commented on Source/Initialization/PlasmaInjector.cpp:
```diff
@@ -563,8 +572,8 @@ void PlasmaInjector::parseMomentum (ParmParse& pp)
     } else if (mom_dist_s == "maxwell_juttner"){
-        Real beta = 0.;
-        Real theta = 10.;
+        Real beta = 0._rt;
+        Real theta = 10._rt;
```
```suggestion
        amrex::Real beta = 0._rt;
        amrex::Real theta = 10._rt;
```

EZoni commented on Source/Initialization/PlasmaInjector.cpp:
```diff
@@ -594,7 +603,7 @@ void PlasmaInjector::parseMomentum (ParmParse& pp)
         h_inj_mom.reset(new InjectorMomentum((InjectorMomentumJuttner*)nullptr, theta, beta, dir));
     } else if (mom_dist_s == "radial_expansion") {
-        Real u_over_r = 0.;
+        Real u_over_r = 0._rt;
```
```suggestion
        amrex::Real u_over_r = 0._rt;
```

EZoni reviewed (changes requested):
Thank you for this PR, Neil! I think it's a great idea to make our inputs more self-explanatory. I have only two comments:

1. Regarding the name `at_rest`, I think I'm fine with it, but some might think that having an "underscored" name in the input could be a bit annoying. It might be worth thinking of single-word alternatives, such as `rest`, `zero`, `null`, and similar ones. I'll let other developers chime in on this.
2. I added a few suggestions to include the prefix `amrex::` in front of the `Real` types used in the code here. It is a small stylistic detail, but I think we decided some time ago that all AMReX types should be prefixed with `amrex::` in WarpX (in fact we wrote it even in our contributing guidelines), so I would suggest to include that prefix in any new code.

NeilZaim commented:
Hi @EZoni and thanks for the review.

Regarding 1., I personally like `at_rest` because it's short and very explicit. But I'm fine with changing this if needed. What do you think is wrong with having underscores? We already have some for this specific input parameter (`maxwell_boltzmann` or `maxwell_juttner`) and for a few other input parameters and we also abundantly use it for the names of the input parameters themselves (e.g. `<species_name>.momentum_distribution_type` in this case). We could consider ignoring the underscores for robustness, which would be easy to do, but it'd probably be a good idea to do this consistently for all input parameters that read a string.

Regarding 2., I agree with you. Usually, I try to always add the `amrex::` prefix, unless when I do small modifications on a file that does not use this convention (mainly for consistency within a file, and also due to laziness). But in this case I modified nearly all the `Real` variables in the file, so it definitely makes sense to add the prefix. I've added it in a few other places in this file.

EZoni commented:
Thank you for adding the `amrex::` prefix consistently, Neil! I think the reason why CI is failing is that there are still a few missing, when we compile with openPMD support. The diff below should tell you which ones and it should fix CI:
```
--- a/Source/Initialization/PlasmaInjector.cpp
+++ b/Source/Initialization/PlasmaInjector.cpp
@@ -310,7 +310,7 @@ PlasmaInjector::PlasmaInjector (int ispecies, const std::string& name)
         queryWithParser(pp_species_name, "z_shift",z_shift);
 
 #ifdef WARPX_USE_OPENPMD
-        if (ParallelDescriptor::IOProcessor()) {
+        if (amrex::ParallelDescriptor::IOProcessor()) {
             m_openpmd_input_series = std::make_unique<openPMD::Series>(
                 str_injection_file, openPMD::Access::READ_ONLY);
 
@@ -375,11 +375,11 @@ PlasmaInjector::PlasmaInjector (int ispecies, const std::string& name)
 
         // Broadcast charge and mass to non-IO processors
         if (!charge_is_specified && !species_is_specified)
-            ParallelDescriptor::Bcast(&charge, 1,
-                ParallelDescriptor::IOProcessorNumber());
+            amrex::ParallelDescriptor::Bcast(&charge, 1,
+                amrex::ParallelDescriptor::IOProcessorNumber());
         if (!mass_is_specified && !species_is_specified)
-            ParallelDescriptor::Bcast(&mass, 1,
-                ParallelDescriptor::IOProcessorNumber());
+            amrex::ParallelDescriptor::Bcast(&mass, 1,
+                amrex::ParallelDescriptor::IOProcessorNumber());
 #else
         amrex::Abort("Plasma injection via external_file requires openPMD support: "
                      "Add USE_OPENPMD=TRUE when compiling WarpX.\n");
```

Regarding the name `at_rest`, I'm totally fine with it. I don't think there's anything wrong with having underscores, I just remember some Zoom discussions in the past where it seemed that some developers/users could have a preference for single-word names over anything else essentially. But I'm personally totally fine with `at_rest`, it's short and self-explanatory.

NeilZaim commented:
Thanks Edoardo for simplifying my work for the OpenPMD compilation. :) Looks like it's working now.
